### PR TITLE
 Milestone 1: minimal TUF

### PR DESCRIPTION
**Pull request by jku**
* included is a client that has matching functionality vs baseline
* repository metadata has been published
* some (repository) design notes have been written

### CODE REVIEW

**Comment by jku** [comment]
A few possible changes still:
* I'm considering making keyids readable names -- it's not strictly spec compliant but it would make this much easier to understand and talk about
* Also considering removal of index.json files: The repository design allows just looking at targetpaths to gather all of this info, it's just that ngclient doesn't currently expose the list of targetpaths (https://github.com/theupdateframework/python-tuf/issues/1995) so this would have to access ngclient internals. The benefit would be a more straightforward implementation that is easier to understand

Both of these could be future enhancements waiting for library improvements as well

**Comment by MVrachev** [comment]
Do you want us to make a high-level review on this @jku or do you want us to wait until you mark it as a non-draft pr?

**Comment by jku** [comment]
> Also considering removal of index.json files

I will try this (by using some ngclient internal interfaces), and see how it works... I will keep this a draft until I see how that works and decide if I implement it in this PR or not

**Comment by jku** [comment]
* I modified the design so index.json is no longer there: this makes both the download process and the repository content simpler
* I don't think there's anything critical missing from calling this milestone 1: 
  * saving artifacts outside of git only makes sense when we have the upload functionality (otherwise it's too much manual work)
  * the way data is encoded in the targetpath could probably be improved... but if there's some really useful invention later, we can always come back to this milestone and improve it (in a branch if needed)
  * the fact that this uses python-tuf internals (to find all the targetpaths) is unfortunate but I think a good way to test the idea for https://github.com/theupdateframework/python-tuf/issues/1995: again, we can always fix this later if python-tuf gets the API (or if we decide that was a dumb idea)

I think this is ready. That said, I welcome opinions on high level decisions as well as the details.

**Comment by jku** [comment]
It would probably be good if the main README documented the current state or milestone at a high level: so people who come to the project for the first time could see both what the state is and what the goal is...

File: docs/design-milestones/01-TUF-MINIMAL.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+* No "server" functionality: Metadata is not modified by any online processes, only by the repository admin
+* No maintenance tools: repository admin is expected to edit metadata using other unspecified methods (https://github.com/vmware-labs/repository-editor-for-tuf was used in practice)
+* repository admin makes metadata changes and signs the changes locally, pushes the changes to a remote git repository: this is automatically published to downloader clients HTTP endpoint
+* A downloader client is provided in this repository, see playground/README.md for client usage.
+
+## Security
+
+The repository admin key is required to add or remove artifacts (in a way that clients would accept). Root key is required to update the repository admin key.
+
+The advantages over baseline repository are:
```
[the remark this answers is not in the document]
Comment (jku) [reply docs/design-milestones/01-TUF-MINIMAL.md:15]: Thanks. I'm a little worried documenting this well gets trickier with more milestones but we'll see.

**Comment by jku** [comment]
Also, something we could already have at milestone 1: hashed bin delegations. This would give a notion of scalability to this setup.

However, I think I'll rather wait for
* succinct delegations to arrive in a python-tuf release
* a bit of repository helper code to arrive in tools (like repository-editor-for-tuf): I'd like a single command to add a file to a repository (so that it actually gets added to the correct targets file). This doesn't currently exist

**Comment by joshuagl** [comment]
Given that we're specifically focused on modelling community content repositories, which tend to be large, including hashed bin delegation early is a good idea.

**Comment by lukpueh** [comment]
This is great! I haven't looked at `playground_client.py` in detail yet, but I very much like the overall direction here. Two questions:

What is the plan for `playground/client` over the course of the roadmap? Now that it has an initial root and a client that supports a minimal TUF repository, I think it should be useable for any repo design of future milestones, bar some minor configuration such as remote `BASE_URL`, and local `METADATA_DIR` to avoid conflicts. Most changes of functionality in `playground/client` (e.g. adopt [theupdateframework/python-tuf#1995](https://github.com/theupdateframework/python-tuf/issues/1995)) should be doable in a backwards-compatible way so that it  works with all repo designs starting at `01-TUF-MINIMAL.md`. I ask this because we briefly talked about a different branch for each milestone, or a single branch with different client directories. But if my assumption above is correct, we won't need that and will instead have a growing `design-milestones/` dir and an evolving `TUF-ROADMAP.md` doc which always captures all accomplished milestones. Also, if this is correct, we could make the `playground/client/README.md` more generic. Of course we can make these decisions as we move along the ROADMAP. If, for instance, we need a custom network implementation for a future client, it could become messy.

The other question concerns the current set of private keys used to sign the metadata. IIUC, these are on your local machine, @jku? Should we share them among administrators on this repo so that they can update any metadata? At least the root key?

**Comment by jku** [comment]
1. client roadmap

> What is the plan for playground/client over the course of the roadmap? Now that it has an initial root and a client that supports a minimal TUF repository, I think it should be useable for any repo design of future milestones...

I fear devil is in the details though: e.g. the current target listing hack depends on knowing the specific targets metadata name. This will break the moment we have any delegations. Likewise I would not be surprised if some unseen requirement means we actually need index.json-like files in the future anyway. So changes are likely to happen even if the core client likely looks roughly the same. Still, no need to create branches until we realize something really should have been in milestone X but wasn't there.

2. keys

Many ways to handle this -- you could just make a PR to the playground-tuf-minimal repository that adds your root key and developer key, and I could sign that change... I can give pointers on easiest ways to do this if this sounds good.

3. succinct delegations

This now works in repository-editor-for-tuf (using unreleased python-tuf) so I want to add a succinct delegation here but I think not in this milestone: it requires solving the target listing issue mentioned above

**Comment by jku** [comment]
I think I will merge this now: let's handle remaining complaints in new issues

**Pull request merged**